### PR TITLE
Feature/crypto

### PR DESCRIPTION
**Pull request by moorts**

### CODE REVIEW

**Comment by moorts** [comment]
Just realized that i probably should move EncryptionInterface to application... thoughts?

File: src/.idea/misc.xml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (erikzmmrmnn) [inline src/.idea/misc.xml:1]: This file should not be included in the commit. I wonder why it is not automatically excluded by the gitignore.

File: src/.idea/src.iml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (erikzmmrmnn) [inline src/.idea/src.iml:1]: This file should not be included in the commit. I wonder why it is not automatically excluded by the gitignore.

File: src/.idea/.gitignore [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Comment (erikzmmrmnn) [inline src/.idea/.gitignore:1]: Is this the actual gitignore? Or is this some generated gitignore by Clion?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -1,7 +1,10 @@
 cmake_minimum_required(VERSION 3.20)
 project(src)
 
+SET(GCC_COVERAGE_LINK_FLAGS    "-lcryptopp")
+SET(CMAKE_EXE_LINKER_FLAGS  "${CMAKE_EXE_LINKER_FLAGS} ${GCC_COVERAGE_LINK_FLAGS}")
+
 set(CMAKE_CXX_STANDARD 20)
 
-add_executable(src main.cpp application/DbInterface.cpp application/DbInterface.h plugins/database/SQLiteDatabase.cpp plugins/database/SQLiteDatabase.h application/AbstractDatabase.h)
+add_executable(src main.cpp encryption/Cipher.h encryption/PasswordEncryptionHandler.h encryption/KeyDerivation.h encryption/Encryption.h encryption/CBC_Cipher.h encryption/BLOCK.h encryption/EncryptionInterface.h application/DbInterface.cpp application/DbInterface.h plugins/database/SQLiteDatabase.cpp plugins/database/SQLiteDatabase.h application/AbstractDatabase.h)
```
Comment (erikzmmrmnn) [inline src/CMakeLists.txt:9]: Since you moved everything to plugins/encryption you also need to adjust the paths here.

File: src/plugins/encryption/CBC_Cipher.h [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+#include <cryptopp/files.h>
+#include <cryptopp/hex.h>
+
+class CBC_Cipher : public Cipher {
+public:
+    CBC_Cipher(BLOCK b) {
+        this->b = b;
+    }
+
+    std::string encrypt(std::string plain, SecByteBlock key) const {
```
Comment (erikzmmrmnn) [inline src/plugins/encryption/CBC_Cipher.h:24]: Why all the code in the header file instead of the cpp file? (same goes for the decrypt method)

File: src/plugins/encryption/Encryption.h [hunk 6]
```diff
@@ -0,0 +1,16 @@
+//
```
Comment (erikzmmrmnn) [inline src/plugins/encryption/Encryption.h:1]: What's this class for? Is it still needed as a placeholder or can it be discarded?

File: src/plugins/encryption/PasswordEncryptionHandler.h [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#include "KeyDerivation.h"
+
+template <class C>
+concept CIPHER = requires(C) {
+    { C::encrypt(NULL, NULL) } -> std::same_as<std::string>;
+    { C::decrypt(NULL, NULL) } -> std::same_as<std::string>;
+};
+
+template <CIPHER C>
+class PasswordEncryptionHandler {
```
Comment (erikzmmrmnn) [inline src/plugins/encryption/PasswordEncryptionHandler.h:17]: What's the difference between this class and EncryptonInterface? It seems to be doing the same only that this class adds the concept. One of those two classes should be enough then, right?

File: src/plugins/encryption/EncryptionInterface.h [hunk 8]
```diff
@@ -0,0 +1,29 @@
+//
+// Created by moorts on 28.10.21.
+//
+
+#ifndef SRC_ENCRYPTIONINTERFACE_H
+#define SRC_ENCRYPTIONINTERFACE_H
+
+
+class EncryptionInterface {
```
Comment (erikzmmrmnn) [inline src/plugins/encryption/EncryptionInterface.h:9]: This class should be moved out of plugins and into the application layer

File: src/plugins/encryption/Cipher.h [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+#include <cryptopp/osrng.h>
+
+#include <iostream>
+
+#ifndef SRC_CIPHER_H
+#define SRC_CIPHER_H
+
+using namespace CryptoPP;
+
+class Cipher {
```
Comment (erikzmmrmnn) [inline src/plugins/encryption/Cipher.h:15]: This class should be moved out of plugins and into the application layer

**Review by erikzmmrmnn — CHANGES_REQUESTED** [review]
Please see comments added to files.
Also: Currently all the functionality is inside the plugin. That should not be the case. With the database plugin it was easy to separate as the functionality was the SQL statement itself. Any ideas on how to extract the functionality here?

File: src/plugins/encryption/Cipher.h (hunk 9 above)
Comment (moorts) [reply src/plugins/encryption/Cipher.h:15]: I'm not so sure thats a good idea, since the implementations of Cipher directly call the libraries functions

Comment (erikzmmrmnn) [reply src/plugins/encryption/Cipher.h:15]: Yes exactly, the implementations do, but this class doesn't. This class is used to inject the dependency into the EncryptionInterface, which correctly noticed by you as well, should be in the application layer. Therefore if this class isn't in the application layer there would be a dependency pointing from the application layer to the plugins layer. However, dependencies should only go from the outside towards the inside and not the other way around.

File: src/application/EncryptionInterface.h [hunk 10]
```diff
@@ -5,6 +5,8 @@
 #ifndef SRC_ENCRYPTIONINTERFACE_H
 #define SRC_ENCRYPTIONINTERFACE_H
 
+#include "../plugins/encryption/KeyDerivation.h"
```
Comment (erikzmmrmnn) [inline src/application/EncryptionInterface.h:8]: This is a dependency which goes from the application towards the plugin. This should not be the case. Dependencies should only point inwards. To fix this make the parameter `master` only and instead call the derive function inside the implementation of encrypt and decrypt. You could also introduce another abstract class for the key derivation but personally I would prefer the first solution.

**Review by erikzmmrmnn — CHANGES_REQUESTED** [review]

File: src/CMakeLists.txt [hunk 11]
```diff
@@ -6,5 +6,5 @@ SET(CMAKE_EXE_LINKER_FLAGS  "${CMAKE_EXE_LINKER_FLAGS} ${GCC_COVERAGE_LINK_FLAGS
 
 set(CMAKE_CXX_STANDARD 20)
 
-add_executable(src main.cpp plugins/encryption/Cipher.h plugins/encryption/KeyDerivation.h plugins/encryption/CBC_Cipher.h plugins/encryption/BLOCK.h application/EncryptionInterface.h application/DbInterface.cpp application/DbInterface.h plugins/database/SQLiteDatabase.cpp plugins/database/SQLiteDatabase.h application/AbstractDatabase.h plugins/encryption/CBC_Cipher.cpp)
+add_executable(src main.cpp application/Cipher.h plugins/encryption/KeyDerivation.h plugins/encryption/CBC_Cipher.h plugins/encryption/BLOCK.h application/DbInterface.cpp application/DbInterface.h plugins/database/SQLiteDatabase.cpp plugins/database/SQLiteDatabase.h application/AbstractDatabase.h plugins/encryption/CBC_Cipher.cpp)
 target_link_libraries(src sqlite3)
```
Comment (erikzmmrmnn) [inline src/CMakeLists.txt:10]: Please add `cryptopp` to the list of linked libraries and instead remove the addition of the compiler flag in lines 4 and 5. I don't know why but it doesn't work for me with the flag, only when it is included with `target_link_libraries`. To be honest it looks more organized, too, in my opinion.

**Review by erikzmmrmnn — CHANGES_REQUESTED** [review]
Looks good in my opinion. Just change the way of including the library please.
As a next step you can write a short paragraph about the crypto library in the plugins section of the Clean Architecture chapter, just like I did for the database.

**Review by erikzmmrmnn — APPROVED** [review]
Looks good
